rfcs/text/0000-query-by-metadata-schema-versions.md

```diff
@@ -0,0 +1,237 @@
+test design](https://docs.google.com/document/d/18C_Wbrr4Huc8yxQXSdNbzeGk0glBjkfq0MGnozwlIB8/edit) has been proposed. 
+One part of design is to require downstream components to modify their production subscriptions to the Data Store 
+Service (DSS) such that they only receive notifications about bundles with which they are compatible. This will allow
+wranglers to submit new data to the DSS without having to wait for compatibility of the whole system and also allows
+downstream components to receive data independent of whether another component is ready as well. Due to the fact that
+there is currently a metadata schema freeze disallowing any changes, major, minor, or patch, to be pushed to production,
+this document simply designs the minimum system required to unblock wranglers from submitting data and protecting
+downstream components (and therefore the entire DCP) from breaking due to incompatibility.
+
+Note: the subscription modifications will only exist in the production environment. Once the metadata schema integration
+tests are setup, subscriptions will be modified to remove any restrictions on metadata schema versions in order to
+eagerly surface incompatibilities.
+
+
+### User Stories
+
+*Share the [User Stories](https://www.mountaingoatsoftware.com/agile/user-stories) motivating this RFC.*
+* As a data wrangle, I would like to be able to push new data adhering to the latest metadata schema without having to
+wait for every single component to be compatible with the new schema versions.
+* As a DCP developer working on a downstream component (e.g. Matrix Service, Secondary Analysis, etc.) I would like to
+receive notification about new data adhering to new metadata schema versions as soon as I am ready without having to
+wait for other DCP components to also be compatible. 
+
+## Detailed Design
+
+There are two questions fueling the design of this system:
+
+1) How will the schema version be noted in the metadata schema?
+2) Will ElasticSearch, JMESPath, and Query Service all be able to issue range queries against the format of the schema
+version?
+
+Today, the version of the metadata schema is embedded in the `described_by` field which contains the schema version
+embedded in the URL; an example value is “https://schema.humancellatlas.org/type/biomaterial/15.3.0/donor_organism” and
+the schema version is 15.3.0. Unfortunately embedding the version within the URL does not facilitate searching for a
+range of data between schema versions. 
+
+### New schema version fields in metadata schema
+We propose adding two new required fields to each list of schema properties:
+
+```
+"schema_major_version": {
+      "description": "The major version number of the schema.",
+      "type": "integer",
+      "pattern": "^[1-9]\\d*$",
+      "example": "4"
+},
+"schema_minor_version": {
+      "description": "The minor version number of the schema.",
+      "type": "integer",
+      "pattern": "^\\d+$",
+      "example": "6"
+}
```

Comment:
I'd like to propose a third field in addition to `schema_major_version` and `schema_minor_version`: `schema_revision_number`. This could be any increasing function of schema changes. For instance, the number of commits to [metadata-schema](https://github.com/HumanCellAtlas/metadata-schema) would probably work, which can be found with the command `git rev-list --count HEAD`.

This would provide an easy way to pin a large swath of metadata.

Reply:
Would the `schema_revision_number` only increment for the files that get changed (so the number only has significance within the context of an individual file)? Or would it function more like a snapshot of all the files (kind of like a commit hash)?

Reply:
@samanehsan More like a snapshot of all of the files.

Reply:
Thanks for the clarification! That would be incredibly valuable to us since analysis could just pin to a particular snapshot.

Reply:
I strongly urge the author to consider the addition of a monotonically increasing revision number as @xbrianh suggests. If we postpone this decision, we'd have to rewrite the metadata again once it is adopted.

The `git rev-list --count HEAD` command could produce the same number for two distinct branches with the same number of commits but with different commits in the divergent parts of the those branches. Adding a forth property, `schema_branch`, could be used to disambiguate them.

Reply:
Who would populate the `schema_revision_number` and `schema_branch` fields? Ingest for primary submissions; data processing pipelines for secondary submissions?

Reply:
@hannes-ucsc I totally understand your perspective and I agree with you that there are a lot of benefits to having an overall version but I cannot agree to adding this at this moment. There are pros and cons to an overall schema which are much more nuanced that this comment thread, not to mention that this is not a decision I can make in a vacuum, and I'm doing my absolute best to consider this option by a) having received permission from PMs/POs to schedule a Spike *in Q3* (https://app.zenhub.com/workspaces/dcp-5ac7bcf9465cb172b77760d9/issues/humancellatlas/dcp/465), b) scheduling the meeting next week with the key stakeholders, and c) committing to writing a pre-RFC document on the pros/cons of having an overall schema version AND potentially ideas on how to compute this/store this and its implications.

I have spent so many hours trying to get to a first MVP that *covers a lot, but not all* use cases and this is finally one that almost everyone has agreed to run with. To your point about having to rewrite metadata again, this is somewhat moot to me once we get migrations rolling since that is what will happen on a regular basis anyways for data releases.

I am asking you to, please, meet me halfway and agree to at least storing the major/minor schema version for now knowing that I have, and I will continue, put in a ton of time trying to expedite the conversation about an overall schema.

Reply:
Could Ingest do it for both?

But thinking more about this, especially adding the branch reference seems difficult to do. Still, the idea of a monotonically increasing revision number, even if not derived from the # of commits in the git branch would be useful. The revision number could be put into schemas when they are published on `schema.humancellatlas.org` and copied into each document by ingest when a document is created.

Or maybe the revision number would just be present in one schema/document pair only: `project.json` has always had a special role. It is currently not linked into the graph and its mere presence in a bundle associates all other documents in that bundle with the project. The revision number could be placed only into the project schema and only copied into `project.json` when that file is created. Client would then have the choice to pin the semantic version of individual documents or the monotonically increasing revision number on `project.json` only. Or a range of either.

Reply:
@hannes-ucsc Pardon, but what do you mean by "for both"?

I don't disagree with you at all that supporting a monotonically increasing revision number is useful. I agree with you! I also think you bring some really great ideas for implementing this to the table and I am definitely going to add all of the comments above to the document I am preparing for next week's meeting.

The only thing I am asking is to at least let the ball start rolling with this proposed MVP. I am 100% committing to iterating on this design quickly.

Reply:
If we allow scope creep we may never be ready. _Release then iterate_ makes sense to me.

Reply:
Yeah, frankly there are already situations where we need to go back and change older bundles (fixing bugs, inserting entities) and I don't see that changing. We can do this iteratively.

A cheap way of doing it on your end might be to concat major and minor version number, as long as it isn't possible for there to be branches where an older schema version still gets a minor version update after there's some separate new major version. But if this was the case a monotonic revision wouldn't be of use anyway.

Reply:
> @hannes-ucsc Pardon, but what do you mean by "for both"?

For both primary and secondary bundles.

> The only thing I am asking is to at least let the ball start rolling with this proposed MVP. I am 100% committing to iterating on this design quickly.

Sounds good. I'll rest my case on this issue. 

Punting this does mean that Analysis has to pin essentially every schema referenced here:

https://github.com/HumanCellAtlas/metadata-api/blob/11480262abe484facebee016ce7e63a366c8743a/src/humancellatlas/data/metadata/api.py#L792

@justincc 

> A cheap way of doing it on your end might be to concat major and minor version number

Wouldn't 1.10 and 11.0 would produce the same value? `int(major) * 100 + int(minor)` might be an alternative.

Anyways, not only does the revision number need to be monotonically increasing, we need one revision number for the entire schema!

Reply:
@hannes-ucsc Apologies  - I'm an idiot and I wrote that late at night. I realized in the car that my proposal wouldn't work.